### PR TITLE
[Improvement][C++] Add validation for data types for writers in C++ library

### PR DESCRIPTION
**Pull request by lixueclaire**
## Proposed changes

This update includes the implementation of various validation levels in the GraphAr C++ SDK writers. These new validation levels could ensure the consistency of data types between the data and meta info before being written.

## Types of changes

What types of changes does your code introduce to GraphAr?
_Put an `x` in the boxes that apply_

- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation Update (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [x] I have read the [CONTRIBUTING](https://github.com/alibaba/GraphAr/blob/main/CONTRIBUTING.rst) doc
- [x] I have signed the CLA
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)

## Further comments

related to issue #127

### CODE REVIEW

File: cpp/src/arrow_chunk_writer.cc [hunk 1]
```diff
@@ -83,15 +83,30 @@ Status VertexPropertyWriter::Validate(
...
+  if (validate_level == ValidateLevel::strong_validate) {
+    auto schema = input_table->schema();
+    for (auto& property : property_group.GetProperties()) {
+      int indice = schema->GetFieldIndex(property.name);
+      if (indice == -1)
+        return Status::InvalidOperation("property: " + property.name +
+                                        " not found");
+      auto field = schema->field(indice);
+      if (DataType::ArrowDataTypeToDataType(field->type()) != property.type)
+        return Status::TypeError("invalid data type for property: " +
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:106]: We can add more error informations like
```
std::string err_msg = "Invalid data type for property: " + property.name + " define as " + property.type.ToTypeName() + " but got " + DataType::ArrowDataTypeToDataType(field->type()).TypeName;
return Status::TypeError(err_msg)
```

File: cpp/src/arrow_chunk_writer.cc [hunk 2]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
     ValidateLevel validate_level) const noexcept {
+  // use the writer's validate level
+  if (validate_level == ValidateLevel::default_validate)
+    validate_level = validate_level_;
+  // no validate
+  if (validate_level == ValidateLevel::no_validate)
+    return Status::OK();
+  // weak validate
+  if (!edge_info_.ContainAdjList(adj_list_type_))
+    return Status::InvalidOperation("invalid adj list type");
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:201]: Ditto, error msg should be " The adj list type: xxx is not exist in edge info"

File: cpp/src/arrow_chunk_writer.cc [hunk 3]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+  if (adj_list_type_ != AdjListType::ordered_by_source &&
+      adj_list_type_ != AdjListType::ordered_by_dest)
+    return Status::InvalidOperation("invalid adj list type");
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:204]: Ditto, "The adj list type has to be ordered_by_source or ordered_by_dest, but got xxx"

File: cpp/src/arrow_chunk_writer.cc [hunk 4]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+  if (adj_list_type_ == AdjListType::ordered_by_source &&
+      input_table->num_rows() > edge_info_.GetSrcChunkSize() + 1)
+    return Status::OutOfRange();
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:207]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 5]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+  if (adj_list_type_ == AdjListType::ordered_by_dest &&
+      input_table->num_rows() > edge_info_.GetDstChunkSize() + 1)
+    return Status::OutOfRange();
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:210]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 6]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+  if (vertex_chunk_index < 0)
+    return Status::InvalidOperation("invalid vertex chunk index");
+  // strong validate
+  if (validate_level == ValidateLevel::strong_validate) {
+    auto schema = input_table->schema();
+    int index = schema->GetFieldIndex(GeneralParams::kOffsetCol);
+    if (index == -1)
+      return Status::InvalidOperation("offsets not provided");
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:218]: "offset column not provided"

File: cpp/src/arrow_chunk_writer.cc [hunk 7]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+    auto field = schema->field(index);
+    if (field->type()->id() != arrow::Type::INT64)
+      return Status::TypeError("invalid data type for offsets");
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:221]: "The data type of offset offset column should be INT64, but got "

File: cpp/src/arrow_chunk_writer.cc [hunk 8]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+      return Status::InvalidOperation("sources not provided");
+    auto field = schema->field(index);
+    if (field->type()->id() != arrow::Type::INT64)
+      return Status::TypeError("invalid data type for sources");
+    index = schema->GetFieldIndex(GeneralParams::kDstIndexCol);
+    if (index == -1)
+      return Status::InvalidOperation("destinations not provided");
+    field = schema->field(index);
+    if (field->type()->id() != arrow::Type::INT64)
+      return Status::TypeError("invalid data type for destinations");
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:258]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 9]
```diff
@@ -175,16 +190,111 @@ Status VertexPropertyWriter::WriteTable(
...
+  if (validate_level == ValidateLevel::strong_validate) {
+    auto schema = input_table->schema();
+    for (auto& property : property_group.GetProperties()) {
+      int indice = schema->GetFieldIndex(property.name);
+      if (indice == -1)
+        return Status::InvalidOperation("property: " + property.name +
+                                        " not found");
+      auto field = schema->field(indice);
+      if (DataType::ArrowDataTypeToDataType(field->type()) != property.type)
+        return Status::TypeError("invalid data type for property: " +
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:294]: Ditto

File: cpp/src/arrow_chunk_writer.cc [hunk 10]
```diff
@@ -231,16 +341,24 @@ Status EdgeChunkWriter::WriteOffsetChunk(
...
   GAR_RETURN_NOT_OK(Validate(input_table, vertex_chunk_index));
   GAR_ASSIGN_OR_RAISE(auto file_type, edge_info_.GetFileType(adj_list_type_));
+  std::vector<int> indices;
+  indices.clear();
+  auto schema = input_table->schema();
+  int index = schema->GetFieldIndex(GeneralParams::kOffsetCol);
+  if (index == -1)
+    return Status::InvalidOperation("offsets not provided");
+  indices.push_back(index);
+  auto in_table = input_table->SelectColumns(indices).ValueOrDie();
```
Comment (acezen) [inline cpp/src/arrow_chunk_writer.cc:351]: Make it simple and elegance

```c++
GAR_ASSIGN_OR_RAISE(in_table, input_table->SelectColumns({index});
```

**Comment by lixueclaire** [comment]
@acezen, thanks for your suggestions, I have added more error informations when validating.

**Review by acezen — APPROVED** [review]
LGTM

**Pull request merged**